torch/distributed/__init__.py

```diff
@@ -54,6 +56,35 @@ def is_available() -> bool:
         _make_nccl_premul_sum,
     )
 
+    def breakpoint(rank: int = 0):
+        """
+        Set a breakpoint, but only on a single rank.  All other ranks will wait for you to be
+        done with the breakpoint before continuing.  This calls ``breakpoint()`` under the
+        hood, so you can customize it using the normal facilities, e.g., ``PYTHONBREAKPOINT``
+        environment variable.
+
+        Args:
+            rank (int): Which rank to break on.  Default: ``0``
+        """
+        if get_rank() == rank:
+            # This will be the case when your subprocess was created by
+            # multiprocessing.Process, see
+            # https://stackoverflow.com/questions/30134297/python-multiprocessing-stdin-input
+            old_stdin = None
+            if isinstance(sys.stdin, io.TextIOWrapper):
+                old_stdin = sys.stdin
+                sys.stdin = open(0)
+            try:
+                breakpoint(header=(
+                    "\n!!! ATTENTION !!!\n\n"
+                    f"Type 'up' to get to the frame that called dist.breakpoint(rank={rank})\n"
+                ))  # type: ignore[call-arg]
```

Comment:
this code doesn't work.

I tried to use it with:

```
    dist.breakpoint(0)
```
and it fails to call `breakpoint()` because `pdb.breakpoint` has no arg `header`

Did you mean to write:

```
           try:
                pdb.set_trace(header=(
                    "\n!!! ATTENTION !!!\n\n"
                    f"Type 'up' to get to the frame that called dist.breakpoint(rank={rank})\n"
                ))  # type: ignore[call-arg]
```
instead?

And then it doesn't set the breakpoint correctly, but not much can be done afterwards. You can inspect the stack and vars, but you can't step from there even if you back out from this frame.

When you step it calls 

```
            finally:
                if old_stdin is not None:
                    sys.stdin.close()
                    sys.stdin = old_stdin
```

and everything breaks because IO gets closed.

```
    dist.breakpoint(0)
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/site-packages/torch/distributed/__init__.py", line 85, in breakpoint
    sys.stdin = old_stdin
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/site-packages/torch/distributed/__init__.py", line 85, in breakpoint
    sys.stdin = old_stdin
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/bdb.py", line 88, in trace_dispatch
    return self.dispatch_line(frame)
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/bdb.py", line 112, in dispatch_line
    self.user_line(frame)
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/pdb.py", line 262, in user_line
    self.interaction(frame, None)
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/pdb.py", line 357, in interaction
    self._cmdloop()
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/pdb.py", line 322, in _cmdloop
    self.cmdloop()
  File "/home/stas/anaconda3/envs/py39-pt22/lib/python3.9/cmd.py", line 126, in cmdloop
    line = input(self.prompt)
ValueError: I/O operation on closed file.
```

Would love to have this working...

Reply:
i fixed this yesterday, hopefully, pls lmk if the fixed version actually works for you!
https://github.com/pytorch/pytorch/pull/115705